analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ConditionEvaluatesToConstantBase.cs

```diff
@@ -63,21 +77,102 @@ public override void ExecutionCompleted()
         }
-        foreach (var operation in alwaysFalse)
+        foreach (var isFalse in alwaysFalse)
         {
-            ReportIssue(operation, false);
+            ReportIssue(isFalse.Key, isFalse.Value, false);
         }
 
         base.ExecutionCompleted();
     }
 
-    private void ReportIssue(IOperation operation, bool conditionEvaluation)
+    private void ReportIssue(IOperation operation, BasicBlock block, bool conditionEvaluation)
     {
         var issueMessage = operation.Kind == OperationKindEx.IsNull ? MessageNullCoalescing : string.Format(MessageBool, conditionEvaluation);
-        ReportIssue(Rule2589, operation, null, issueMessage);
+        var secondaryLocations = SecondaryLocations(block, conditionEvaluation);
+        if (secondaryLocations.Any())
+        {
+            ReportIssue(Rule2583, operation, secondaryLocations, string.Format(MessageUnreachable, issueMessage));
+        }
+        else
+        {
+            ReportIssue(Rule2589, operation, null, string.Format(issueMessage, conditionEvaluation));
+        }
     }
+
+    private List<Location> SecondaryLocations(BasicBlock block, bool conditionIsTrue)
+    {
+        var unreachable = Unreachable(block, conditionIsTrue).Where(x => IsNotIgnored(x.Syntax));
+        List<Location> locations = new();
+        IOperation currentStart = null;
+        IOperation currentEnd = null;
+        foreach (var operation in unreachable.Union(reached).OrderBy(x => x.Syntax.SpanStart))
+        {
+            if (reached.Contains(operation))
+            {
+                if (currentStart is not null)
+                {
+                    locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+                    currentStart = null;
+                }
+            }
+            else
+            {
+                currentStart ??= operation;
+                currentEnd = operation;
+            }
+        }
+        if (currentStart != null)
+        {
+            locations.Add(currentStart.Syntax.CreateLocation(currentEnd.Syntax));
+        }
+        return locations;
+    }
+
+    private static IEnumerable<IOperation> Unreachable(BasicBlock block, bool conditionIsTrue)
+    {
+        if (block.SuccessorBlocks.Distinct().Count() != 2)
+        {
+            return Enumerable.Empty<IOperation>();
+        }
+        HashSet<BasicBlock> reachable = new() { block };
+        HashSet<BasicBlock> unreachable = new();
+        var successor = conditionIsTrue ^ block.ConditionKind == ControlFlowConditionKind.WhenFalse
+            ? block.ConditionalSuccessor.Destination
+            : block.SuccessorBlocks.Single(x => x != block.ConditionalSuccessor.Destination);
+        TraverseReachable(successor);
+        var unreachableSuccessor = block.SuccessorBlocks.Single(x => x != successor);
+        TraverseUnreachable(unreachableSuccessor);
+        return unreachable.SelectMany(x => x.OperationsAndBranchValue);
```

Comment:
```suggestion
        return unreachable.SelectMany(x => x.OperationsAndBranchValue).Except(reached);
```
As discussed - let's return the unreachable operations from all reachable blocks.